File: 05 - SOLID Principles/Single Responsibility Principle/Java Code/ViolatesSRP.java
// =============================================================================
// SINGLE RESPONSIBILITY PRINCIPLE — the VIOLATION (Java twin of violates-srp.cpp).
//
// Order does FOUR unrelated jobs: total, persistence, receipt formatting and
// email. Four reasons to change live in one class.
//
// Build & run:
//   javac ViolatesSRP.java
//   java ViolatesSRP
// =============================================================================
import java.util.ArrayList;
import java.util.List;

public class ViolatesSRP {

    static class Line {
        String item;
        double price;
        int qty;
        Line(String item, double price, int qty) {
            this.item = item;
            this.price = price;
            this.qty = qty;
        }
    }

    // ONE "god class" with four responsibilities glued together.
    static class Order {
        List<Line> lines = new ArrayList<Line>();

        double total() {
            double t = 0;
            for (Line l : this.lines) {
                t += l.price * l.qty;
            }
            return t;
        }

        void saveToDatabase() {
            System.out.println("  [DB] INSERT INTO orders ... total = Rs " + (long) total());
        }

        String formatReceipt() {
            StringBuilder s = new StringBuilder("  RECEIPT\n");
            for (Line l : this.lines) {
                s.append("    ").append(l.item).append(" x").append(l.qty).append("\n");
            }
            s.append("    TOTAL: Rs ").append((long) total());
            return s.toString();
        }

        void emailReceipt() {
            System.out.println("  [EMAIL] sending receipt:\n" + formatReceipt());
        }
    }

    public static void main(String[] args) {
        Order order = new Order();
        order.lines.add(new Line("USB-C Cable", 299, 2));
        order.lines.add(new Line("Phone Case", 499, 1));

        order.saveToDatabase();
        order.emailReceipt();
    }
}

File: 05 - SOLID Principles/Single Responsibility Principle/C++ Code/violates-srp.cpp
// =============================================================================
// SINGLE RESPONSIBILITY PRINCIPLE — the VIOLATION.
// "A class should have only ONE reason to change."
//
// Below, Order does FOUR unrelated jobs at once:
//   (1) computes the total, (2) saves itself to a database,
//   (3) formats a printed receipt, and (4) emails the customer.
// Four reasons to change are welded into one class — editing the email format
// risks breaking total calculation, and none of it can be tested in isolation.
//
// Build & run:
//   g++ -std=c++14 violates-srp.cpp -o violates
//   ./violates
// =============================================================================
#include <iostream>
#include <string>
#include <vector>

using namespace std;

struct Line {
    string item;
    double price;
    int qty;
};

// ONE "god class" with four responsibilities glued together.
class Order {
public:
    vector<Line> lines;

    // Responsibility 1: pricing rules.
    double total() {
        double t = 0;
        for (int i = 0; i < (int)this->lines.size(); i++) {
            t += this->lines[i].price * this->lines[i].qty;
        }
        return t;
    }

    // Responsibility 2: persistence (changes whenever the database changes).
    void saveToDatabase() {
        cout << "  [DB] INSERT INTO orders ... total = Rs " << (long)this->total() << endl;
    }

    // Responsibility 3: presentation (changes whenever the receipt layout does).
    string formatReceipt() {
        string s = "  RECEIPT\n";
        for (int i = 0; i < (int)this->lines.size(); i++) {
            s += "    " + this->lines[i].item + " x" + to_string(this->lines[i].qty) + "\n";
        }
        s += "    TOTAL: Rs " + to_string((long)this->total());
        return s;
    }

    // Responsibility 4: notification (changes whenever the email provider does).
    void emailReceipt() {
        cout << "  [EMAIL] sending receipt:\n" << this->formatReceipt() << endl;
    }
};

int main() {
    Order* order = new Order();
    order->lines.push_back({"USB-C Cable", 299, 2});
    order->lines.push_back({"Phone Case", 499, 1});

    // One object doing everything: pricing, DB, formatting AND email.
    order->saveToDatabase();
    order->emailReceipt();

    delete order;
    return 0;
}
